### DISCUSSION TITLE
Type 1 fonts

### DISCUSSION THREAD

**Discussion by mattiasmarka** [olsak/OpTeX#104] [Q&A] [answered]
Hi

Is there anyway to use Type 1 fonts? The documentation doesn't mention Type 1 fonts at all (nor do the TUGboat articles), so I can't quite tell whether Type 1 fonts are supported or not.

The documentation uses the term "Unicode fonts". Am I to understand this means OpenType and TrueType fonts only? I don't know a whole lot about font encodings, so it would be great if someone could clarify.

Thanks

**Comment by olsak** [comment]
The \font primitive supports tfm files and LuaTeX supports pdftex.map, \pdfmapline etc. If you prepare your font files (tfm, enc, mapline, pfb, maybe vf) and install them to appropriate places (like when pdftex is used), then  you can load such font by \font and use it. Moreover, you have to solve a re-encoding from Unicode to the font encoding of the font: the example is in OpTeX trick  http://petr.olsak.net/optex/optex-tricks.html#nonunicode .

I strongly recommend not to bother with these old font technology which needs many files, complicated installation and they don't support Unicode directly.

**Reply by mattiasmarka** [reply]
Thanks.

I'm using the Lucida family, which comes in both Type 1 and OpenType formats. Unfortunately some fonts of the family aren't available in the OpenType format. For example. there is no OpenType version of the seriffed typewriter font.

Follow-up question: If I manage do get Type 1 fonts working, will font modifiers like `\caps` work by default or do I have to do some special setup?

**Comment by olsak** [comment, marked as answer]
The `\caps` font modifier (declared in font family files) is based on OpenType font feature. If we want to implement something similar with Type1 fonts then we have to create a macro which selects different tfm file where Caps is present. It is not supported by default, unfortunately. And if we want to mix OpenType fonts and Type1 in single family and create font modifiers for such a family, then this is more complicated task.

You can do something more simple. If you want to use these fonts only in a single document, use the `\font` primitive with directly declared sizes for all Type1 fonts you need and select these fonts by font selectors declared by the `\font`. Or, you can use `\sfont` instead `\font`. It declares a macro which behaves like font selector but moreover, it keeps the current font size too.

**Comment by mattiasmarka** [comment]
I observed that the following works:
```
\initunifonts
\font\myfont=file:lbr at10pt
\myfont test with unicode letter äöü
```
Now, this isn't documented behavior. @olsak, is this safe to use? That is, can the user assume this works?

**Comment by olsak** [comment]
It is documented: section 2.12: Using \font primitive.

To your particular question. I don't know what is the `lbr` file. If it is a tfm file then it is encoded according to some encoding. Typical is T1 encoding in TeX. Then most letters from latin1 encoding set (Western European countries) will work because they are coded at the same slot as Unicode. But not all, for example German ß is encoded differently in T1 and in Unicode.

**Reply by mattiasmarka** [reply]
Sorry, I forgot `lbr` isn't part of a regular TeX installation. I'll use a different example.

**Reply by mattiasmarka** [reply]
Hmm... I still don't see it in the documentation. Perhaps we're talking about different things.

Let me be more specific.

Consider:
```
\font\myfont=ArevSans-Roman.afm at10pt
\myfont test äöü ß
\bye
```
This produces an error: "metric data not found or bad".
But consider:
```
\initunifonts
\font\myfont=ArevSans-Roman.afm at10pt
\myfont test äöü ß
\bye
```
This works fine.

Is it safe to assume the last example always works in OpTeX?

ArevSans.afm is NOT a Unicode font, right? So, why does `\initunifonts` make it accessible? Does the documentation say this should be the case?

Thanks for being so patient.

**Reply by vlasakm** [reply]
I can't try your examples right now, and you are doing something very unusual, so hopefully I am not too wrong.

Usually when people use Type 1 fonts with TeX they use the original `.pfb` fonts (which contains the glyph drawings) and a `.tfm` file (TeX font metrics) which is specially generated from the original `.afm` file (Adobe fonts metrics). This is because normal TeX can only ever read its own TeX font metrics format (and in theory doesn't care at all about the drawings of the glyphs, just their dimensions). This is what @olsak talks about and this is also the misunderstanding - you are trying to use `.afm` files directly, not `.tfm` files.

While not built into TeX, LuaTeX can be through Lua code extended to understand other font formats. In particular the Lua code which is loaded by `\initunifonts` (and which allows OpenType fonts) can also handle Type 1 font metrics directly without going through `.tfm` files. This is nice, because IIRC it removes the 256 character limit and the problem of encoding vectors, map files and other baggage that normaly needs to be present for use of Type 1 fonts in TeX. 

IIRC the ConTeXt Lua code that does this supports `.afm` files by essentially translating them as if they were OpenType, so this means that you still have a lot of flexibility even with Type 1 fonts. See [section 1.5 Type1 of fonts-mkiv.pdf](https://texdoc.org/serve/fonts-mkiv/0)  for more details. They also recommend to use Type 1 fonts with `.afm` (if available) instead of `.tfm`.

Now this is very unusual because AFAIK nobody uses this outside of ConTeXt. [Luaotfload ](https://texdoc.org/serve/luaotfload/0)documents it a little bit (occurences of `afm`) and in OpTeX it essentially falls under the category of OpenType fonts - an extension through Lua code.

From my point of view: if it works, use it. AFAIK there are no plans to remove these features. You are already aware that OpenType is much better if available and won't use this unnecessarily.

Though you can also go through the classical `.tfm` route and there are much more resources on that topic.

**Reply by mattiasmarka** [reply]
That clears things up --- thanks.

**Comment by olsak** [comment]
OK, I'll add a short note about `\font\f=file.afm` to my OpTeX documentation (in similar sense as in the section 1.5 of mentioned "Fonts out of ConTeXt").
